src/quantuminspire/projectq/backend_qx.py

```diff
@@ -167,65 +299,63 @@ def _store(self, cmd: Command) -> None:
             self._measured_ids += [logical_id]
+            # do not add the measurement statement when fsp is possible
             if not self._full_state_projection:
-                self.qasm += "\nmeasure q[{}]".format(cmd.qubits[0][0].id)
+                self.qasm += "\nmeasure q[{}]".format(sim_qubit_id)
             return
 
         # when we find a gate after measurements we don't have fsp
         # add any delayed measurement statements
-        if self._full_state_projection and (len(self._measured_ids) != 0):
-            for logical_qubit_id in self._measured_ids:
-                physical_qubit_id = self._logical_to_physical(logical_qubit_id)
-                self.qasm += "\nmeasure q[{}]".format(physical_qubit_id)
-            self._full_state_projection = False
+        if self._full_state_projection and len(self._measured_ids) != 0:
+            self._switch_fsp_to_nonfsp()
 
         if gate == NOT and get_control_count(cmd) == 1:
             # this case also covers the CX controlled gate
-            ctrl_pos = cmd.control_qubits[0].id
-            qb_pos = cmd.qubits[0][0].id
+            ctrl_pos = self._physical_to_simulated(cmd.control_qubits[0].id)
+            qb_pos = self._physical_to_simulated(cmd.qubits[0][0].id)
             self.qasm += "\ncnot q[{}], q[{}]".format(ctrl_pos, qb_pos)
         elif gate == Swap:
-            q0 = cmd.qubits[0][0].id
-            q1 = cmd.qubits[1][0].id
+            q0 = self._physical_to_simulated(cmd.qubits[0][0].id)
+            q1 = self._physical_to_simulated(cmd.qubits[1][0].id)
             self.qasm += "\nswap q[{}], q[{}]".format(q0, q1)
         elif gate == X and get_control_count(cmd) == 2:
-            ctrl_pos1 = cmd.control_qubits[0].id
-            ctrl_pos2 = cmd.control_qubits[1].id
-            qb_pos = cmd.qubits[0][0].id
+            ctrl_pos1 = self._physical_to_simulated(cmd.control_qubits[0].id)
+            ctrl_pos2 = self._physical_to_simulated(cmd.control_qubits[1].id)
+            qb_pos = self._physical_to_simulated(cmd.qubits[0][0].id)
             self.qasm += "\ntoffoli q[{}], q[{}], q[{}]".format(ctrl_pos1, ctrl_pos2, qb_pos)
         elif gate == Z and get_control_count(cmd) == 1:
-            ctrl_pos = cmd.control_qubits[0].id
-            qb_pos = cmd.qubits[0][0].id
+            ctrl_pos = self._physical_to_simulated(cmd.control_qubits[0].id)
+            qb_pos = self._physical_to_simulated(cmd.qubits[0][0].id)
             self.qasm += "\ncz q[{}], q[{}]".format(ctrl_pos, qb_pos)
         elif gate == Barrier:
-            qb_pos = [qb.id for qr in cmd.qubits for qb in qr]
+            qb_pos_list = [qb.id for qr in cmd.qubits for qb in qr]
             self.qasm += "\n# barrier gate "
             qb_str = ""
-            for pos in qb_pos:
-                qb_str += "q[{}], ".format(pos)
+            for pos in qb_pos_list:
+                qb_str += "q[{}], ".format(self._physical_to_simulated(pos))
```

Comment:
Not requesting changes, but just for fun then two lines above can be written as:
qb_str = `''.join(map(lambda x: f'p[{self._physical_to_simulated(x)}], ', qb_pos_list ))`
Don't think it's much cleaner, but always fun to make one-liners

Reply:
or just:
`''.join([f'p[{self._physical_to_simulated(x)}], ' for x in qb_pos_list ])`

Reply:
Nice, but you mean q[...] I suppose?
Maybe even better is making use of the separator string ', ':
', '.join([f'q[{self._physical_to_simulated(x)}]' for x in qb_pos_list])
In the end I could half the number of lines to create the barrier qasm string.